group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -0,0 +1,1009 @@
+     */
+    public void scheduleLoadOperation(
+        TopicPartition tp,
+        int partitionEpoch
+    ) {
+        log.info("Scheduling loading of metadata from {} with epoch {}", tp, partitionEpoch);
+        getOrCreateContext(tp);
+
+        scheduleInternalOperation("Load(tp=" + tp + ", epoch=" + partitionEpoch + ")", tp, () -> {
+            CoordinatorContext context = getOrCreateContext(tp);
+            if (context.epoch < partitionEpoch) {
+                context.epoch = partitionEpoch;
+
+                switch (context.state) {
+                    case INITIAL:
+                    case FAILED:
+                    case CLOSED:
+                        context.transitionTo(CoordinatorState.LOADING);
+                        loader.load(tp, context.coordinator).whenComplete((state, exception) -> {
+                            scheduleInternalOperation("CompleteLoad(tp=" + tp + ", epoch=" + partitionEpoch + ")", tp, () -> {
+                                CoordinatorContext ctx = contextOrThrow(tp);
+                                try {
+                                    if (exception != null) throw exception;
+                                    ctx.transitionTo(CoordinatorState.ACTIVE);
+                                    log.info("Finished loading of metadata from {} with epoch {}.", tp, partitionEpoch);
+                                } catch (Throwable ex) {
+                                    log.error("Failed to load metadata from {} due to {}.", tp, ex);
+                                    ctx.transitionTo(CoordinatorState.FAILED);
+                                }
+                            });
+                        });
+                        break;
+
+                    case LOADING:
+                        log.info("The coordinator {} is already loading metadata.", tp);
+                        break;
+
+                    case ACTIVE:
+                        log.info("The coordinator {} is already active.", tp);
+                        break;
+                }
+            } else {
+                log.info("Ignoring loading metadata from {} since current epoch {} is larger than or equals to {}.",
+                    context.tp, context.epoch, partitionEpoch);
+            }
+        });
+    }
+
+    /**
+     * Schedules the unloading of a coordinator. This is called when the broker is not the
+     * leader anymore.
+     *
+     * @param tp                The topic partition of the coordinator.
+     * @param partitionEpoch    The partition epoch.
+     */
+    public void scheduleUnloadOperation(
+        TopicPartition tp,
+        int partitionEpoch
+    ) {
+        log.info("Scheduling unloading of metadata for {} with epoch {}", tp, partitionEpoch);
+
+        scheduleInternalOperation("UnloadCoordinator(tp=" + tp + ", epoch=" + partitionEpoch + ")", tp, () -> {
+            CoordinatorContext context = contextOrThrow(tp);
+            if (context.epoch < partitionEpoch) {
```

Comment:
Does the context.epoch stay the same until the next load?

Reply:
Right. The epoch is basically the leader epoch of the partition. The broker notifies us when a new ones comes and we update it here. However, we only load if we have to.